### PR TITLE
dont hardcode sockets per node

### LINKED ISSUE #97: don't hardcode sockets_per_node in openmpi.py
either figure out what sockets_per_node is or make it an option?
https://github.com/hpcugent/vsc-mympirun/blob/develop/lib/vsc/mympirun/mpi/openmpi.py#L69

### PR DESCRIPTION
**Pull request by Caylo**
fix #97

### CODE REVIEW

File: lib/vsc/mympirun/mpi/openmpi.py [hunk 1]
```diff
@@ -28,6 +28,7 @@
 Documentation can be found at https://www.open-mpi.org/doc/
 """
 import os
+import subprocess
```
Comment (boegel) [inline lib/vsc/mympirun/mpi/openmpi.py]: this is not being used?

File: lib/vsc/mympirun/mpi/openmpi.py [hunk 2]
```diff
@@ -66,7 +67,20 @@ def pinning_override(self):
         self.log.debug("pinning_override: type %s ", override_type)
 
         ranktxt = ""
-        sockets_per_node = 2
+
+        sockets_per_node = self.options.sockets
+        if not sockets_per_node:
```
Comment (boegel) [inline lib/vsc/mympirun/mpi/openmpi.py]:
```python
if sockets_per_node == 0:
```

File: lib/vsc/mympirun/option.py [hunk 3]
```diff
@@ -123,8 +123,7 @@ def make_init(self):
 
             "showsched": ("Print the known Sched classes and exit", None, "store_true", False, 's'),
 
-
-            "socket": ("Force socket device", None, "store_true", None),
+            "sockets-per-node": ("Number of sockets per cpu (default/0: get from /proc/cpuinfo)", "int", "store", 0),
```
Comment (boegel) [inline lib/vsc/mympirun/option.py]: `s/per cpu/per node/g` (a CPU is in a socket ;))

also, clarify w.r.t. default: `(default: 0, i.e. try to detect # sockets via /proc/cpuinfo)`

File: lib/vsc/mympirun/mpi/openmpi.py [hunk 4]
```diff
@@ -66,7 +67,20 @@ def pinning_override(self):
...
+            try:
+                proc_cpuinfo = open('/proc/cpuinfo').read()
+            except IOError as err:
+                self.log.debug("No info on sockets found; hardcoding to 2 (most likely)")
+                sockets_per_node = 2
+
+            if proc_cpuinfo:
+                res = re.findall('^physical id.*', proc_cpuinfo, re.M)
+                sockets_per_node = len(nub(res))
+                self.log.debug("Sockets per node found in cpuinfo: set to %s" % sockets_per_node)
```
Comment (boegel) [inline lib/vsc/mympirun/mpi/openmpi.py]: maybe add a failsafe, if `sockets_per_node` is 0, something went very wrong... throw an error in the case, and instruct to specify `--sockets-per-node`

File: lib/vsc/mympirun/mpi/openmpi.py [hunk 5]
```diff
@@ -66,7 +67,20 @@ def pinning_override(self):
...
         ranktxt = ""
-        sockets_per_node = 2
+
+        sockets_per_node = self.options.sockets
+        if not sockets_per_node:
+            try:
+                proc_cpuinfo = open('/proc/cpuinfo').read()
+            except IOError as err:
+                self.log.debug("No info on sockets found; hardcoding to 2 (most likely)")
+                sockets_per_node = 2
```
Comment (boegel) [inline lib/vsc/mympirun/mpi/openmpi.py]: better fail hard here, and mention to use `--sockets-per-node`?

File: lib/vsc/mympirun/mpi/openmpi.py [hunk 6]
```diff
@@ -66,7 +69,23 @@ def pinning_override(self):
...
+                self.log.error("Number of sockets per node not found: please use --sockets-per-node")
+                sys.exit(1)
+
+            if proc_cpuinfo:
+                res = re.findall('^physical id.*', proc_cpuinfo, re.M)
+                sockets_per_node = len(nub(res))
+                if sockets_per_node == 0:
+                    self.log.error("Something went wrong: found sockets per node = 0. Please use --sockets-per-node")
+                    sys.exit(1)
+                self.log.debug("Sockets per node found in cpuinfo: set to %s" % sockets_per_node)
```
Comment (boegel) [inline lib/vsc/mympirun/mpi/openmpi.py]: @Caylo let's put this in a separate method, this `pinning_override` method is getting too big with this included

```python
def determine_sockets_per_node(self):
   """Useful docstring here"""
   ...
   return sockets_per_node

def pinning_override(self):
    ...
    sockets_per_node = self.determine_sockets_per_node()
```

Can we come up with a way to test this btw?

**Comment by Caylo** [comment]
Jenkins proc/cpuinfo:
```
processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 15
model name	: Intel(R) Xeon(R) CPU E5-2690 v2 @ 3.00GHz
stepping	: 1
microcode	: 0x416
cpu MHz		: 3000.000
cache size	: 25600 KB
fpu		: yes
fpu_exception	: yes
cpuid level	: 13
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss syscall nx lm constant_tsc arch_perfmon pebs bts nopl xtopology tsc_reliable nonstop_tsc aperfmperf pni ssse3 cx16 hypervisor lahf_lm ida arat xsaveopt pln pts dtherm
bogomips	: 6000.00
clflush size	: 64
cache_alignment	: 64
address sizes	: 40 bits physical, 48 bits virtual
power management:

processor	: 1
vendor_id	: GenuineIntel
cpu family	: 6
model		: 15
model name	: Intel(R) Xeon(R) CPU E5-2690 v2 @ 3.00GHz
stepping	: 1
microcode	: 0x416
cpu MHz		: 3000.000
cache size	: 25600 KB
fpu		: yes
fpu_exception	: yes
cpuid level	: 13
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss syscall nx lm constant_tsc arch_perfmon pebs bts nopl xtopology tsc_reliable nonstop_tsc aperfmperf pni ssse3 cx16 hypervisor lahf_lm ida arat xsaveopt pln pts dtherm
bogomips	: 6000.00
clflush size	: 64
cache_alignment	: 64
address sizes	: 40 bits physical, 48 bits virtual
power management:

processor	: 2
vendor_id	: GenuineIntel
cpu family	: 6
model		: 15
model name	: Intel(R) Xeon(R) CPU E5-2690 v2 @ 3.00GHz
stepping	: 1
microcode	: 0x416
cpu MHz		: 3000.000
cache size	: 25600 KB
fpu		: yes
fpu_exception	: yes
cpuid level	: 13
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss syscall nx lm constant_tsc arch_perfmon pebs bts nopl xtopology tsc_reliable nonstop_tsc aperfmperf pni ssse3 cx16 hypervisor lahf_lm ida arat xsaveopt pln pts dtherm
bogomips	: 6000.00
clflush size	: 64
cache_alignment	: 64
address sizes	: 40 bits physical, 48 bits virtual
power management:

processor	: 3
vendor_id	: GenuineIntel
cpu family	: 6
model		: 15
model name	: Intel(R) Xeon(R) CPU E5-2690 v2 @ 3.00GHz
stepping	: 1
microcode	: 0x416
cpu MHz		: 3000.000
cache size	: 25600 KB
fpu		: yes
fpu_exception	: yes
cpuid level	: 13
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss syscall nx lm constant_tsc arch_perfmon pebs bts nopl xtopology tsc_reliable nonstop_tsc aperfmperf pni ssse3 cx16 hypervisor lahf_lm ida arat xsaveopt pln pts dtherm
bogomips	: 6000.00
clflush size	: 64
cache_alignment	: 64
address sizes	: 40 bits physical, 48 bits virtual
power management:
```

**Comment by boegel** [comment]
Maybe `lscpu` is a better option than reading `/proc/cpuinfo`, see: https://access.redhat.com/discussions/480953

**Comment by boegel** [comment]
merged via #129

**Pull request merged**